### PR TITLE
Don't use totality when defining the (potentially) partial order relation <=

### PR DESCRIPTION
**Pull request by andreasnoack**

### CODE REVIEW

**Review by StefanKarpinski — APPROVED** [review]

File: test/operators.jl [hunk 1]
```diff
@@ -97,3 +97,15 @@ end
     B = 3 .> [1 -1 5] .> 0
     @test B == [true false false]
 end
+
+Base.:(<)(x::Type, y::Type) = (x <: y) & (x != y)
```
Comment (yuyichao) [inline test/operators.jl]: I think this should be wrapped in a custom type e.g. `struct SortType x end`. While no other test should be relying on this, there could be future interference should another test want to do something similar.

**Comment by alyst** [comment]
Potentially it's 2 comparisons vs 1. What would nanosoldier say?

**Comment by andreasnoack** [comment]
> Potentially it's 2 comparisons vs 1. What would nanosoldier say?

For integers, the generated code is very similar, but why guess when you can just ask

@nanosoldier `runbenchmarks(ALL, vs = ":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/22027#issuecomment-303465673) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/188baa34bcf3c3e90be14c14ab45f835814d4548/8073503_vs_cfa1999/report.md). cc @jrevels

**Comment by StefanKarpinski** [comment]
This came up on discourse: https://discourse.julialang.org/t/behavior-of-when-isless-or-is-defined/10114. Looking at it again, I'm wondering what the motivation for the change was. As a general fallback since even if this doesn't cause a performance problem for built-in types, user-defined types may have expensive `<` and `==` operations and this always calls them both.

**Comment by andreasnoack** [comment]
The main motivation was correctness. We state that `<` is a partial order relation so it seemed a bit odd to use totality here. I think fallbacks should be correct and special cases optimized. The problem is that it is not easy to communicate that users might benefit from defining special `<=` for their type.

Just thinking out loud a possible alternative: my impression is that the `isless`/`<` thing is mainly there because of floats and not because of some deeper order theory considerations. Hence an alternative could be that `<` should generally be a total order and then clearly state that floating points are special and has a very special `isless` function that are not supposed to be used for anything but floats.

**Comment by StefanKarpinski** [comment]
I do agree that `<` being a partial order is a bit fishy.

**Pull request merged**